feature/dynamodb/enhancedclient/marshaler_test.go

```diff
@@ -0,0 +1,769 @@
+				} else if err.Error() != c.err.Error() {
+					t.Fatalf("marshalList with input %#v returned error `%s`, expected error `%s`",
+						c.input, err, c.err)
+				}
+				return
+			}
+			if err != nil {
+				t.Fatalf("MarshalList with input %#v returned error `%s`", c.input, err)
+			}
+
+			compareObjects(t, c.expected, actual)
+
+		})
+	}
+}
+
+func Test_New_UnmarshalList(t *testing.T) {
+	// Using the same inputs from MarshalList, test the reverse mapping.
+	for name, c := range marshallerListTestInputs {
+		t.Run(name, func(t *testing.T) {
+			if c.input == nil {
+				t.Skip()
+			}
+
+			iv := reflect.ValueOf(c.input)
+
+			actual := reflect.New(iv.Type())
+			if iv.Kind() == reflect.Slice {
+				actual.Elem().Set(reflect.MakeSlice(iv.Type(), iv.Len(), iv.Cap()))
+			}
+
+			if err := UnmarshalList(c.expected.([]types.AttributeValue), actual.Interface()); err != nil {
+				t.Errorf("unmarshal with input %#v returned error `%s`", c.expected, err)
+			}
+			compareObjects(t, c.input, actual.Elem().Interface())
+		})
+	}
+}
+
+func Test_New_UnmarshalListError(t *testing.T) {
+	// Test that we get an error using UnmarshalList to convert to a nil value.
+	expected := &InvalidUnmarshalError{Type: reflect.TypeOf(nil)}
+	if err := UnmarshalList[any](nil, nil); err == nil {
+		t.Errorf("UnmarshalList with input %T returned no error, expected error `%v`", nil, expected)
+	} else if err.Error() != expected.Error() {
+		t.Errorf("UnmarshalList with input %T returned error `%v`, expected error `%v`", nil, err, expected)
+	}
+
+	// Test that we get an error using UnmarshalList to convert to a non-pointer value.
+	var actual map[string]interface{}
+	expected = &InvalidUnmarshalError{Type: reflect.TypeOf(actual)}
+	if err := UnmarshalList(nil, actual); err == nil {
+		t.Errorf("UnmarshalList with input %T returned no error, expected error `%v`", actual, expected)
+	} else if err.Error() != expected.Error() {
+		t.Errorf("UnmarshalList with input %T returned error `%v`, expected error `%v`", actual, err, expected)
+	}
+
+	// Test that we get an error using UnmarshalList to convert to nil struct.
+	var actual2 *struct{ A int }
+	expected = &InvalidUnmarshalError{Type: reflect.TypeOf(actual2)}
+	if err := UnmarshalList(nil, actual2); err == nil {
+		t.Errorf("UnmarshalList with input %T returned no error, expected error `%v`", actual2, expected)
+	} else if err.Error() != expected.Error() {
+		t.Errorf("UnmarshalList with input %T returned error `%v`, expected error `%v`", actual2, err, expected)
+	}
+}
+
+func compareObjects(t *testing.T, expected interface{}, actual interface{}) {
+	t.Helper()
+	if !reflect.DeepEqual(expected, actual) {
+		ev := reflect.ValueOf(expected)
+		av := reflect.ValueOf(actual)
+		if diff := cmpDiff(expected, actual); len(diff) != 0 {
+			t.Errorf("expect kind(%s, %T) match input kind(%s, %T)\n%s",
+				ev.Kind(), ev.Interface(), av.Kind(), av.Interface(), diff)
+		}
+	}
+}
+
+func BenchmarkMarshalOneMember(b *testing.B) {
```

Comment:
thanks for writing those. I'm not sure if we usually run Go stdlib benchmark tests, curious if you found something interesting running them or have any initial numbers